File: Study Plan/Top Interview 150/0392 Is Subsequence/IsSubsequence.java
/*  
    Problem
    --------------------------------------------------------------------------------
    392. Is Subsequence
*/

/* 
    Solution
    --------------------------------------------------------------------------------
    - Runtime 1 ms (Beats 91.07% of users with Java)
    - Memory 41.24 MB (Beats 74.55% of users with Java)
*/

class Solution {
    public boolean isSubsequence(String s, String t) {
        if (s.length() == 0) return true;
        String subSequence = s;
        int subSequenceLength = subSequence.length();
        int subSequencePointer = 0;
        String sequence = t;
        int sequenceLength = sequence.length();
        int sequencePointer = 0;
        while (subSequencePointer < subSequenceLength
            && sequencePointer < sequenceLength) {
            if (subSequence.charAt(subSequencePointer) == sequence.charAt(sequencePointer)) {
                subSequencePointer++;
                sequencePointer++;
            } else {
                sequencePointer++;
            }
        }
        return subSequencePointer == subSequenceLength;
    }
}

File: Study Plan/Top Interview 150/0392 Is Subsequence/IsSubsequence.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    392. Is Subsequence
*/

/* 
    Solution
    --------------------------------------------------------------------------------
    - Runtime 54 ms (Beats 68.78% of users with C#)
    - Memory 39.92 MB (Beats 31.94% of users with C#)
*/

public class Solution
{
    public bool IsSubsequence(string s, string t)
    {
        if (string.IsNullOrWhiteSpace(s)) return true;

        var subSequence = s;
        var subSequenceLength = subSequence.Length;
        var subSequencePointer = 0;

        var sequence = t;
        var sequenceLength = sequence.Length;
        var sequencePointer = 0;

        while (subSequencePointer < subSequenceLength
            && sequencePointer < sequenceLength)
        {
            if (subSequence[subSequencePointer] == sequence[sequencePointer])
            {
                subSequencePointer++;
                sequencePointer++;
            }
            else
            {
                sequencePointer++;
            }
        }

        return subSequencePointer == subSequenceLength;
    }
}